### PR TITLE
Add type guards 

### PR DESCRIPTION
**Pull request by dtnhinh**
- [x] Related GitHub issue(s) linked in PR description
Isssue #73 
- [x] Destination branch merged, built and tested with your changes
- [x] Code formatted and follows best practices and patterns
- [x] Code builds cleanly (no _additional_ warnings or errors)
- [x] Manually tested
- [x] Automated tests are passing
- [x] No _decreases_ in automated test coverage
- [x] Documentation updated (readme, docs, comments, etc.)
- [x] Localization: No hard-coded error messages in code files (_minimally_ in string constants)

### CODE REVIEW

**Comment by dtnhinh** [comment]
Working on resolving the conflict :/

File: src/utilities/collection-utils.ts [hunk 1]
```diff
@@ -330,22 +333,22 @@ const _take = <T>(array: Array<T> | null | undefined, n?: number): T[] =>
...
-    intersection: _intersection,
-    intersectionWith: _intersectionWith,
-    length: _length,
-    removeElementAt: _removeElementAt,
-    replaceElementAt: _replaceElementAt,
-    sample: _sample,
-    sampleSize: _sampleSize,
-    sortByString: _sortByString,
-    take: _take,
+    difference,
```
Comment (wintondeshong) [inline src/utilities/collection-utils.ts:339]: 👍

File: src/utilities/string-utils.ts [hunk 2]
```diff
@@ -28,7 +28,7 @@ const filename = (value?: string): string | undefined =>
  * @param {string} [value]
  * @returns {boolean}
  */
-const hasValue = (value?: string): boolean =>
+const hasValue = (value?: string): value is string =>
```
Comment (wintondeshong) [inline src/utilities/string-utils.ts:31]: Nice 👍

File: src/utilities/collection-utils.test.ts [hunk 3]
```diff
@@ -320,18 +229,9 @@ describe("CollectionUtils", () => {
...
-            const nullCollection = (null as unknown) as any[];
-            const undefinedCollection = (undefined as unknown) as any[];
-
-            // Act
-            const result = CollectionUtils.isNotEmpty(
-                undefinedCollection,
-                nullCollection,
-                collection
-            );
+        test("when collections param is an array with element, it returns true", (): void => {
```
Comment (wintondeshong) [inline src/utilities/collection-utils.test.ts:232]: Appreciate you updating this and other tests. Very thorough job!

**Review by wintondeshong — APPROVED** [review]
@nhinhdao thanks! nice work

**Comment by dtnhinh** [comment]
Thanks, @wintondeshong. Please don't merge in just yet. There are some conflicts I am not understanding why they are still here

**Review by brandongregoryscott — APPROVED** [review]

**Comment by brandongregoryscott** [comment]
Thanks for working through this, @nhinhdao! And thank you @myty for bringing up the idea 👏

**Comment by brandongregoryscott** [comment]
@dtnhinh While it's not failing any tests, or the build, I'm getting a TS error from the `isEmpty` type guard where it's being used in `StringUtils.join`:

image

I am going to update the type guard to be `collection is undefined` instead of `collection is any[] | Immutable.List<any> | undefined`. I think for the purposes of using that function, `undefined`, `null`, or `size/count === 0` can be considered the same.

And the same for `hasValues` - the type guard should really be `collection is any[] | Immutable.List<any>`, not `collection is any[] | Immutable.List<any> | undefined`. We don't gain any benefit from the type guard if TS still thinks it is nullable.

image
image

Sorry I didn't catch these during the PR, but I'll make these changes before the publish.

**Comment by dtnhinh** [comment]
I didn't see the join error on my end so not sure if it is Rider thing or not.

For the `collection is any[] | Immutable.List<any> | undefined`, I totally agree with you on that. Giving anything `| undefined` defeats the purpose of the type guard. I didn't realize that lol.

Thank you for pointing it out and fixing that @brandongregoryscott

**Pull request merged**